File: routes/v1/poweron_container.go
package v1

import (
	"fmt"
	"os"
	"os/exec"
	"server/method"
	"server/method/db"
	"strings"

	"github.com/gofiber/fiber/v2"
)

func PowerOnContainer(app *fiber.App) {

	app.Get("/api/v1/instance/poweron/:bot_id/:admin_key/", func(c *fiber.Ctx) error {

		var bot_id = c.Params("bot_id")
		var admin_key = c.Params("admin_key")

		if !method.ValidateAdminKey(admin_key) {
			fmt.Println("[PowerOn] Erreur admin_key", admin_key, " n'est pas valide!")
			return c.Status(fiber.StatusBadRequest).SendString("Invalid admin_key!")
		}

		if _, err := os.Stat(method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id)); os.IsNotExist(err) {
			fmt.Println("[PowerOn] Erreur bot_id", bot_id, " n'existe pas!")
			return c.Status(fiber.StatusBadRequest).SendString("Invalid bot_id!")
		}

		var IsOn = db.Get(bot_id + "_online")

		if !IsOn {
			cliArray := []struct {
				L   string
				CWD string
			}{
				{
					L:   "npx tsc",
					CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id),
				},

				{
					L:   strings.Replace("mv dist/index.js dist/{Code}.js", "{Code}", bot_id, 1),
					CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id),
				},

				{
					L:   strings.Replace("pm2 start dist/{Code}.js -f", "{Code}", bot_id, 1),
					CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id),
				},
			}

			for _, index := range cliArray {
				cmd := exec.Command("sh", "-c", index.L)
				cmd.Dir = index.CWD
				cmd.Stdout = os.Stdout
				cmd.Stderr = os.Stderr
				if err := cmd.Run(); err != nil {
					fmt.Print(err)
					return c.Status(fiber.StatusInternalServerError).SendString(err.Error())
				}
			}

		} else {
			fmt.Println("[PowerOn] Erreur tentative doublon!")
		}

		db.Set(bot_id+"_online", true)
		return c.SendStatus(200)
	})

}

File: src/Routes/v1/poweron_container.ts
import { validateAdminKey } from '../../method/validateData.js';
import isContainerOn from '../../method/getContainerStatus.js';
import config from '../../method/getConfigData.js';

import { Request, Response } from 'express';
import fs from "node:fs";
import { execSync } from "child_process";
import path from "node:path";
import { db } from '../../method/database.js';
import { getOwnerByCode } from '../../method/getOwnerByCode.js';
import logsRequest from '../../method/logRequest.js';

export const route = {
    type: 'get',
    apiPath: '/api/v1/instance/poweron/:bot_id/:admin_key/',
    run: async (req: Request, res: Response) => {
        logsRequest(route, req);

        let ownihrz_table = db.table("OWNIHRZ");

        const botId = req.params["bot_id"];
        const adminKey = req.params["admin_key"];

        if (!config?.api.apiToken) {
            console.log("Error: Failed to load config");
            return res.status(500).send("Failed to load config");
        }

        if (!validateAdminKey(adminKey)) {
            console.log("[Delete] Erreur admin_key n'est pas valide!");
            return res.status(403).send("Invalid admin_key!");
        };

        if (!fs.existsSync(path.join(process.cwd(), 'ownihrz', botId))) {
            console.log("[Delete] Erreur bot_id n'existe pas!");
            return res.status(403).send("Invalid bot_id!");
        };

        if (!await isContainerOn(botId)) {
            [
                {
                    line: `npx tsc`,
                    cwd: path.join(process.cwd(), 'ownihrz', botId)
                },
                {
                    line: `mv dist/index.js dist/${botId}.js`,
                    cwd: path.join(process.cwd(), 'ownihrz', botId)
                },
                {
                    line: `pm2 start ./dist/${botId}.js -f`,
                    cwd: path.join(process.cwd(), 'ownihrz', botId)
                },
            ].forEach((index) => {
                try {
                    execSync(index.line, { stdio: [0, 1, 2], cwd: index.cwd });
                } catch (e: any) {
                    console.log(e.toString().split('\n')[0]);
                }
            });
        } else {
            console.log('[Startup] Erreur tentative doublon!');
        }

        let ownerid1 = getOwnerByCode(await ownihrz_table.get("CLUSTER"), botId);
        await ownihrz_table.set(`CLUSTER.${ownerid1}.${botId}.PowerOff`, false);

        return res.sendStatus(200);
    },
};